### PR TITLE
Hs 334 fc

### PR DESCRIPTION
**Pull request by David-T-White**
Fix Fibre Channel support with FC initiator discovery and passing. Update initiator address handling code to distinguish between SAS and FC

### CODE REVIEW

File: pkg/node/node.go [hunk 1]
```diff
@@ -121,6 +122,11 @@ func (node *Node) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest)
 		topology[topoKey] = sasAddr
 	}
 
+	for i, sasAddr := range fcAddresses {
+		topoKey := fmt.Sprintf("%s/%s-%d", common.TopologyInitiatorPrefix, common.TopologyFCInitiatorLabel, i)
+		topology[topoKey] = sasAddr
+	}
```
Comment (jskazinski) [inline pkg/node/node.go:128]:
```suggestion
	for i, fcAddr := range fcAddresses {
		topoKey := fmt.Sprintf("%s/%s-%d", common.TopologyInitiatorPrefix, common.TopologyFCInitiatorLabel, i)
		topology[topoKey] = fcAddr
	}
```

Comment (jskazinski) [reply pkg/node/node.go:128]: In the fmt.Sprintf("%s/%s-%d", common.TopologyInitiatorPrefix, common.TopologyFCInitiatorLabel, i), you are printing the index, nit the fcAddr, is that the intent?

File: pkg/storage/fcNode.go [hunk 2]
```diff
@@ -276,3 +283,69 @@ func (fc *fcStorage) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCa
...
+		} else {
+			klog.Infof("found FC initiator address %s", address)
+			discoveredFCAddresses = append(discoveredFCAddresses, address)
+		}
+	}
+	return discoveredFCAddresses, nil
+}
+
+// Read the fc address configuration file and return addresses
+func readFCAddrFile() ([]string, error) {
```
Comment (jskazinski) [inline pkg/storage/fcNode.go:329]: Suggest passing in filename string instead of hard coding it here.

File: pkg/storage/fcNode.go [hunk 3]
```diff
@@ -276,3 +283,69 @@ func (fc *fcStorage) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCa
...
+
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			} else {
+				klog.Warningf("error searching for FC HBA addresses: %v", err)
+				return nil, err
+			}
+		} else {
+			klog.Infof("found FC initiator address %s", address)
```
Comment (jskazinski) [inline pkg/storage/fcNode.go:321]: I suggest we start switching to structured logging for new log messages ( don't use Infof() ):
			klog.Info("found FC initiator", "address", address)

File: pkg/storage/sasNode.go [hunk 4]
```diff
@@ -53,6 +53,75 @@ func (sas *sasStorage) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnsta
 	return nil, status.Error(codes.Unimplemented, "NodeUnstageVolume is not implemented")
 }
 
+// Read the sas address configuration file and return addresses
+func readSASAddrFile() ([]string, error) {
```
Comment (jskazinski) [inline pkg/storage/sasNode.go:57]: Consider passing in filename string to provide a bit more flexibility.

File: pkg/storage/sasNode.go [hunk 5]
```diff
@@ -53,6 +53,75 @@ func (sas *sasStorage) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnsta
...
+
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			} else {
+				klog.Warningf("error searching for SAS HBA addresses: %v", err)
+				return nil, err
+			}
+		} else {
+			klog.Infof("found SAS HBA address %s", address)
```
Comment (jskazinski) [inline pkg/storage/sasNode.go:113]: Suggest structured logging:
			klog.Info("found SAS HBA", "address", address)

File: pkg/storage/sasNode.go [hunk 6]
```diff
@@ -53,6 +53,75 @@ func (sas *sasStorage) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnsta
...
+		sasAddrFile := filepath.Join(scsiHostBasePath, hostDir.Name(), sasAddrFilename)
+		addrBytes, err := os.ReadFile(sasAddrFile)
+		address := string(addrBytes)
+		address = strings.TrimLeft(strings.TrimRight(address, "\n"), "0x")
+
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			} else {
+				klog.Warningf("error searching for SAS HBA addresses: %v", err)
```
Comment (jskazinski) [inline pkg/storage/sasNode.go:109]: If this is an error, suggest using, relying on the passed in filename value:
klog.Error(err, "searching for SAS HBA addresses", "filename", filename)

**Review by jskazinski — APPROVED** [review]
Some suggested changes, with one that may require code changes in pkg/node/node.go, but the others are suggestions.

File: pkg/node/node.go (hunk 1 above)
Comment (David-T-White) [reply pkg/node/node.go]: Yes that's intentional, the address is passed as the value, the index is added only to make the map key unique

File: pkg/storage/fcNode.go (hunk 3 above)
Comment (David-T-White) [reply pkg/storage/fcNode.go]: Switched to structured logging methods (InfoS, WarningS, ErrorS) in the most recent version that I pushed

File: pkg/storage/fcNode.go (hunk 2 above)
Comment (David-T-White) [reply pkg/storage/fcNode.go]: Passing filename as suggested in updated version

**Review by seagate-chris — APPROVED** [review]
Looks good.

**Pull request merged**